strutil/set.go

```diff
@@ -0,0 +1,96 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package strutil
+
+// OrderedSet is a set of strings that maintains the order of insertion.
+//
+// External synchronization is required for safe concurrent access.
+type OrderedSet struct {
+	positionOf map[string]int
+}
+
+// Items returns a slice of strings representing insertion order.
+//
+// Contains is O(N) in the size of the set.
+func (o *OrderedSet) Items() []string {
+	if o.positionOf == nil {
+		return nil
+	}
+	items := make([]string, len(o.positionOf))
+	for item, idx := range o.positionOf {
+		items[idx] = item
+	}
+	return items
+}
+
+// Contains returns true if the set contains a given item.
+//
+// Contains is O(1) in the size of the set.
+func (o *OrderedSet) Contains(item string) bool {
+	if o.positionOf == nil {
+		return false
+	}
+
+	_, ok := o.positionOf[item]
+	return ok
+}
+
+// IndexOf returns the position of an item in the set.
+func (o *OrderedSet) IndexOf(item string) (idx int, ok bool) {
+	idx, ok = o.positionOf[item]
+	return idx, ok
+}
+
+// Del removes an item from the set.
+//
+// Delete is O(1) in the size of the set.
+func (o *OrderedSet) Del(item string) {
+	if o.positionOf == nil {
+		return
+	}
+
+	delete(o.positionOf, item)
+	if len(o.positionOf) == 0 {
+		o.positionOf = nil
+	}
+}
+
+// Put adds an item into the set.
+//
+// If the item was not present then it is stored and ordered after all existing
+// elements. If the item was already present its position is not changed.
+//
+// Put is O(1) in the size of the set.
+func (o *OrderedSet) Put(item string) {
+	if o.positionOf == nil {
+		o.positionOf = make(map[string]int)
+	}
+
+	if _, ok := o.positionOf[item]; ok {
+		return
+	}
+
+	o.positionOf[item] = len(o.positionOf)
+}
+
+// Size returns the size of the set.
```

Comment:
This doc is bad and you should feel bad.
What is the size of the set? The number of elements that it currently contains? The number of elements that have been added? The sum of the lengths of all the strings it contains? The number of bytes allocated to it? To them?

Peeking at the implementation, why `Size` and not `Len` or `Length`?

Reply:
Size is more of a "number of elements" whereas Length is "the physical dimension of something", at least to me. I don't know if there is a commonly used term for this in Go, for beyond-builtin types.

As for the doc string, how about
```
// Size returns the cardinality of the set
```

or `number of elements` instead

Reply:
cardinality is very mathy and specific so +1 from me as a nerd, but I don't know how non-nerds would read that, I guess probably first thing they would do is google "cardinality" and find the definition, but perhaps we should try to write docs that don't require folks to google first? how about something simple and obvious like:

```go
// Size returns the number of elements currently in the set
```

(possibly without `currently` I don't know if that's clear)

Reply:
```golang
func (s OrderedSet) Cardinality() bool
...
s.Cardinality(🐦) == true
```

Reply:
Done